quinn-proto/src/new_token_store.rs

```diff
@@ -0,0 +1,228 @@
+//! Storing tokens sent from servers in NEW_TOKEN frames and using them in subsequent connections
+
+use bytes::Bytes;
+use slab::Slab;
+use std::{
+    collections::{hash_map, HashMap},
+    mem::take,
+    rc::Rc,
+    sync::Mutex,
+};
+
+/// Responsible for storing tokens sent from servers in NEW_TOKEN frames and retreiving them for
+/// use in subsequent connections
+pub trait NewTokenStore: Send + Sync {
+    /// Called when a NEW_TOKEN frame is received from the server
+    fn store(&self, server_name: &str, token: Bytes);
+
+    /// Called when trying to connect to a server
+    ///
+    /// The same token should never be returned from `take` twice.
+    fn take(&self, server_name: &str) -> Option<Bytes>;
+}
+
+/// `NewTokenStore` implementation that stores up to `N` tokens per server name for up to a limited
+/// number of server names, in-memory
+pub struct InMemNewTokenStore<const N: usize>(Mutex<InMemNewTokenStoreState<N>>);
+
+impl<const N: usize> InMemNewTokenStore<N> {
+    /// Construct empty
+    pub fn new(size_limit: usize) -> Self {
+        Self(Mutex::new(InMemNewTokenStoreState::new(size_limit)))
+    }
+}
+
+impl<const N: usize> NewTokenStore for InMemNewTokenStore<N> {
+    fn store(&self, server_name: &str, token: Bytes) {
+        self.0.lock().unwrap().store(server_name, token)
+    }
+
+    fn take(&self, server_name: &str) -> Option<Bytes> {
+        self.0.lock().unwrap().take(server_name)
+    }
+}
+
+/// Defaults to a size limit of 256
+impl<const N: usize> Default for InMemNewTokenStore<N> {
+    fn default() -> Self {
+        Self::new(256)
+    }
+}
+
+// safety: InMemNewTokenStoreState is !Send and !Sync because it uses Rc. so long as
+//         InMemNewTokenStore doesn't keep Rc handles after the mutex is unlocked, it's safe.
+unsafe impl<const N: usize> Send for InMemNewTokenStore<N> {}
+unsafe impl<const N: usize> Sync for InMemNewTokenStore<N> {}
+
+#[derive(Debug)]
+struct InMemNewTokenStoreState<const N: usize> {
+    size_limit: usize,
+    // linked hash table structure
+    lookup: HashMap<Rc<str>, usize>,
+    entries: Slab<InMemNewTokenStoreEntry<N>>,
+    oldest_newest: Option<(usize, usize)>,
+}
+
+#[derive(Debug)]
+struct InMemNewTokenStoreEntry<const N: usize> {
+    server_name: Rc<str>,
+    older: Option<usize>,
+    newer: Option<usize>,
+    // tokens are pushed to and popped from the top of the token stack. exceeding capacity is
+    // handled by dropping the bottom of the stack.
```

Comment:
It's a bit weirder than that. Queue is FIFO, stack is LIFO. This is LIFO for usage, but FIFO for eviction. I guess it could be described as a dequeue, yeah, but I was sort of thinking of eviction as the exceptional case and usage as the essential case.

Reply:
Yeah nevermind I changed it.